### ISSUE TITLE
Broken volume mounting under Windows

### ISSUE THREAD

**Issue by waynew** [docker/docker#18756]
When running Docker (1.9) under Windows 8, attempting to mount a volume fails miserably:

```
$ docker run --entrypoint /bin/sh -it -v "/c/Path To My Repo:/myprogram" my_test
invalid value "C:\\Path To My Test\\;C:\\Program Files\\Git\\myprogram" for flag -v: bad mode specified: \Program Files\Git\myprogram
See 'C:\Program Files\Docker Toolbox\docker.exe run --help'.
```

This is being run from the mingw prompt.

What am I doing wrong here?

**Comment by thaJeztah** [comment]
Does the path you're trying to mount contain a space? Then this may be a duplicate of  https://github.com/docker/docker/issues/18120, which will be solved by https://github.com/docker/docker/pull/18206 once merged.

**Comment by pewsey** [comment]
I have the same issues. Not using paths with spaces.  Windows 10.

This same image works perfectly under MacOS/X (where it was created)

[code]
$ docker run -it joneaves/xxx
Use of uninitialized value $source_realm in open at /app/find_them.pl line 31.
Use of uninitialized value $source_realm in concatenation (.) or string at /app/find_them.pl line 31.
Could not open file '' No such file or directory at /app/find_them.pl line 31.
[/code]

This is the container running "correctly" (and I'm not passing in the volume information)

[code]
docker run -v `pwd`:/cah joneaves/xxx /cah/d10.csv /cah/m10.csv
[/code]

Works fine on MacOS - but under Windows 10 I get

[code]
$ docker run -v `pwd`:/cc joneaves/xxx
invalid value "C:\\Users\\jon\\coding\\cah;C:\\Program Files\\Git\\cc" for flag -v: bad mode specified: \Program Files\Git\cc
See 'C:\Program Files\Docker Toolbox\docker.exe run --help'.
[/code]

Note the rather weird extraneous "C:\\Program Files\\Git" - which confuses me.  So, how about using that winpty command for docker to see what happens

[code]
$ winpty docker run -v `pwd`:/cc joneaves/xxx
invalid value "C:\\Users\\jon\\coding\\cah:\\cc" for flag -v: bad mode specified
: \cc
See 'docker run --help'.
[/code]

Well, at least it's not complaining about Git this time.

Changing to not use `pwd` and actually specify the path doesn't help either.

[code]
$ winpty docker run -v /c/Users/jon/coding/cah:/cc joneaves/xxx
invalid value "C:\\Users\\jon\\coding\\cah:\\cc" for flag -v: bad mode specified
: \cc
See 'docker run --help'.
[/code]

All of this is done using a terminal created from the Docker Quickstart on Windows

[code]
$ docker version
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.3
 Git commit:   a34a1d5
 Built:        Fri Nov 20 17:56:04 UTC 2015
 OS/Arch:      windows/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.3
 Git commit:   a34a1d5
 Built:        Fri Nov 20 17:56:04 UTC 2015
 OS/Arch:      linux/amd64
[/code]

**Comment by thaJeztah** [comment]
@pewsey thanks for the detailed repro steps! I don't have access to a windows machine, but wondering if adding a double slash helps;

    winpty docker run -v //c/Users/jon/coding/cah://cc joneaves/xxx

I think that prevented some "magic" conversion

**Comment by pewsey** [comment]
`
$ winpty docker run -v //c/Users/jon/coding/cah://cc joneaves/xxx
invalid value "\\\\c\\Users\\jon\\coding\\cah:\\cc" for flag -v: \cc is not an a
bsolute path
See 'docker run --help'.
`

Is the result.

**Comment by pewsey** [comment]
Is there an update on this one? Fixed in a future build? Never to be fixed?

**Comment by legrandjeremy** [comment]
Hi all,

I'm running Windows 10 and I can't use mount volume with docker. I have no error displayed but all sync folders are empty.

**Comment by thaJeztah** [comment]
ping @jhowardmsft is this something that needs a fix in docker/docker?

**Comment by jhowardmsft** [comment]
At a Linux or Windows daemon? No idea about the winpty thing, I just run from native cmd or Powershell, not from bash and winpty that might inappropriate munge and escape paths (such as the double back-slashes in the example). If pointing at a Windows daemon, it needs to be in Windows semantics. eg docker run -v c:\path:d:. If pointing at a Linux daemon, it needs to be in Linux semantics eg docker run -v /foo:/bar. All this assumes you are running a daemon and client with the volume PR merged (such as TP4 or later, can't remember exact date of merge, but something like Nov 23rd was TP4 final, I think you're close on Nov 20th).

**Comment by pewsey** [comment]
@jhowardmsft thanks for that.  You can see my attempts above. Can you craft me a working command line from your perspective.  It would be great if you could try it using the Windows "Docker Quickstart".

I'm not 100% convinced it's not me doing something insane - but I have tried a few permutations which I've documented above and it's not behaving like I'd expect. 

If there's anything else I can do to help this - please feel free to reply to me here. I'd really like to know how to get this to work.

**Comment by jhowardmsft** [comment]
I don't have a "Docker Quickstart", never used it. Just an Ubuntu VM on Hyper-V running docker. Pointing at it from a Windows box: Top window shows the directory contents on the Linux box. Blue windows is on Windows starting a container with a bind mount of /foo into /bar in the container, and subsequently showing the same contents.
![volume](https://cloud.githubusercontent.com/assets/10522484/12301494/31f146b6-b9d5-11e5-8187-f8135591c2fc.JPG)

**Comment by waynew** [comment]
@jhowardmsft That's not the problem that we're having here. Docker running on an ubuntu VM under Windows isn't quite the same as installing the Docker Toolbox on Windows ;)

Which is where the problem is. Your docker client is just talking to the docker daemon inside your VM, mounting a linux path.

*our* problem is trying to mount a Windows path, like this:

```
Windows Machine
+-------------------------------+
|                               |
|  Docker VM (VirtualBox)       |
|  +-----------------------+    |
|  |                       |    |
|  | +------------------+  |    |
|  | | Docker Container |  |    |
|  | |                  |  |    |
|  | | /foo --,         |  |    |
|  | +--------|---------+  |    |
|  |        __|            |    |
|  |       /               |    |
|  |     ,/                |    |
|  |     V                 |    |
|  | /c/Path to/foo -,     |    |
|  +-----------------|-----+    |
|                   /           |
| C:\Path to\foo <-'            |
+-------------------------------+

```

But it doesn't work. Miserably so :(

**Comment by jhowardmsft** [comment]
Ah, AFAIK you can't do that from either a Linux or a Windows client. The volume is on the host running the daemon, not on the client.

**Comment by waynew** [comment]
@jhowardmsft You can mount any folder that the daemon can see. Macs have a similar layout (only mac, not Windows, of course). On my Mac I have no problem doing the same thing.

The virtual machine that Docker creates mounts the `C:` drive as `/c/` - you can verify this if you install the docker toolbox and run `docker-machine ssh default`, then `ls /c/`. The drives are mounted. You can even see in my original post that Docker translates to the Windows version of the path. And then dies.

**Comment by tiborvass** [comment]
@waynew do you have any logs we can have a look at?

**Comment by tiborvass** [comment]
Oh nvm, it's a client-only issue?

**Comment by waynew** [comment]
I *assume* it's a server issue. I don't actually have a Windows machine at my disposal, but I suspect this is the case for any/all windows machines that try to mount a windows volume in the container.

**Comment by cpuguy83** [comment]
This looks like a client issue, particularly with mflag parsing the options.

**Comment by pewsey** [comment]
```
$ docker version
Client:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:14:21 2016
 OS/Arch:      windows/amd64

Server:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 20:39:58 2016
 OS/Arch:      linux/amd64
```

This issue is resolved for my particular use case. Thank you for whatever happened. I suspect it was as cpuguy83 mentioned - some CLI parsing errors.

**Comment by thaJeztah** [comment]
Thanks @pewsey 

@waynew can you confirm if this is resolved for you as well on 1.10.x?

**Comment by ppisljar** [comment]
i am still having this issue 
- docker toolbox installation
- docker version 1.10.2
- windows 7

actually its behaving really weirdly

i tried this: 

```
$ docker run -ti -v /c/:/c_drive debian:latest bash
root@3cd5ac7d0dc7:/# cd /c_drive/
root@3cd5ac7d0dc7:/c_drive# dir
Program\ Files\ (x86)  Users  js
```

as you can see, some folders are shown ... but most are not ....
- in Users there seems to be everything (folders + all files)
- in program files there is just the docker folder
- in js there is nothing ... but it was not showing before ... only after i tried to mount this specific folder ...

anyway, the problem is the same. when i mount a folder from windows machine ... its empty inside docker container.

however mounting /c/Users works as expected ...

**Comment by thaJeztah** [comment]
@ppisljar I think that is expected; by default, only the `C:\Users` directory is shared with the Virtual Machine through virtual box file sharing; https://github.com/boot2docker/boot2docker/#virtualbox-guest-additions

**Comment by 4gekkman** [comment]
I have same problem as @legrandjeremy

I'm running Windows 8.1 and I can't use mount volume with docker. I have no error displayed but all sync folders are empty.

**Comment by ppisljar** [comment]
i got it working with -v /c/Users/german/volumes:/german (note the capital U in users)

thanks @thaJeztah

**Comment by thaJeztah** [comment]
Good to hear, and that actually makes sense, because the Virtual Machine runs Linux, and uses a case-sensitive filesystem, therefore `/c/Users` and `/c/users` is not the same.

**Comment by 4gekkman** [comment]
I'v found why this doesn't work:
```
              docker run `
                -it `
                --volume /c/users/german/volumes:/german `
                ubuntu:14.04 `
                /bin/bash
```
Because container runs not with -d (as daemon), and after run we immediately have command line to container. And in this point of time, has not yet finished mounting, and that's why we have empty folder.
And it's case insencitive.

But if instead:
```
--volume /c/users/german/volumes:/german
'''
i use:
```
--volume /c/users/german/volumes
```
i still have empty folder in container. Any ideas?

**Comment by bizmate** [comment]
I am running on 1.10.3. Noticed @thaJeztah removed this from 1.10.3 but I am still experiencing the problem. 

For instance with

```
version: '2'

services:
  data:
    container_name: data_mounts
    image: busybox
    volumes:
      - ./logs:/logs
  web:
    container_name: web_proxy
    image: nginx:stable
    ports:
      - 80:80
    volumes_from:
      - data
    #volumes:
    #  - ./docker/config/nginx.conf:/etc/nginx/nginx.conf:ro
```
if I then go on the machine and do ls on /logs the folder is empty

```
$ docker exec -it web_proxy bash
root@1abe18c00b73:/# ls /logs/
root@1abe18c00b73:/#
```

Also I am not sure if related to Windows or the container. If uncommenting the docker nginx.conf volume abote i also get error
```

$ docker-compose up -d
Starting data_mounts
Recreating web_proxy
ERROR: Cannot start container 1d4bacb70be8702f98e77c8faac1e33bd7b67d44f8c4ed1684fbc414cfb7a035: [9] System error: not a directory
```

Is these related to this issue? Any suggestions? Should i open a separate one?

**Comment by gabehesse** [comment]
I'm running version 1.12.0.
`docker run -it -v /c/Users/BAMF/codes/TNT:/app tnt_web /bin/bash`
This command did not give an error or warning, resulted in an empty /app directory in my container.
I used the **little whale icon in the taskbar to go to a settings/shared drives** screen. I checked off the C drive and clicked Apply. It made me enter my windows credentials. I reset docker from the Settings menu after this change. Now the same command resulted in a properly mounted /app directory.

**Comment by bddenhartog** [comment]
I can confirm that the process described by @gabehesse results in a properly mounted volume (located on the drive(s) selected in Docker for Windows client's settings) on a new installation of Windows 10.

**Comment by fehergeri13** [comment]
![image](https://cloud.githubusercontent.com/assets/7180398/18270828/3b385b82-742f-11e6-8c74-eb55d73b0b0a.png)
 This solves the problem

**Comment by bddenhartog** [comment]
To clarify, @fehergeri13 posted a screenshot of the solution provided by @gabehesse.

**Comment by cfstras** [comment]
For a mingw bash shell, this is what docker toolbox adds (for example in bashrc) to stop mingw from corrupting the paths:
```
docker () {
  MSYS_NO_PATHCONV=1 docker.exe "$@"
}
export -f docker
```

**Comment by thaJeztah** [comment]
I'm closing this issue because all issues reported here seem to either be an issue with mingw, docker toolbox / VirtualBox guest additions, or docker for windows; for issues with "docker for windows", please use the dedicated issue tracker: https://github.com/docker/for-win/issues. For docker toolbox, please use https://github.com/docker/toolbox/issues.

This is not a bug in the "docker engine", so there's not a lot we can do in this repository, so I'll close
